### ISSUE TITLE
sticky OT: Thorin's soapbox

### ISSUE THREAD

**Issue by Thorin-Oakenpants** [ghacksuserjs/ghacks-user.js#189]
https://www.techdirt.com/articles/20170803/14480037916/once-again-with-feeling-anonymized-data-isnt-really-anonymous.shtml

> Dewes described some methods by which a canny broker can find an individual in the noise, just from a long list of URLs and timestamps. Some make things very easy: for instance, anyone who visits their own analytics page on Twitter ends up with a URL in their browsing record which contains their Twitter username, and is only visible to them. Find that URL, and you’ve linked the anonymous data to an actual person. A similar trick works for German social networking site Xing.

Can't do much about sites dictating the URL like this. Think of all the sites where you have an account and you're logged in, and for the page to work it includes your userid or username.

Not sure where all this data is coming from, but surely HTTPS means this is obscured from snooping

**Comment by fmarier** [comment]
> Not sure where all this data is coming from

One way is through the `Referer` header.

**Comment by Thorin-Oakenpants** [comment]
hmmm .. so we use default `network.http.referer.trimmingPolicy` as `0` which is the full URI
 - full URI: https://example.com:8888/foo/bar.html?id=1234  <- awesome, referer leaks my ID

we also have `network.http.referer.XOriginTrimmingPolicy` at a default of `0` for X-Origin, again 0 stands for full URI, but we only do this for base domains that match (see `network.http.referer.XOriginPolicy", 1`)

Either way, we can leak identifying info in our current user.js via referer. That said, there is no easy solution/fix. Not going down that discussion path again. What is needed is a granular per domain extension, so the interwebs don't break (eg default deny or spoof all, and override for trouble sites)

I also guess some ISPs sell their data to these data brokers. Wouldn't be surprised if Verizon and Comcast sell access.

**Comment by Thorin-Oakenpants** [comment]
https://trac.torproject.org/projects/tor/ticket/23104 - CSS line-height reveals the platform Tor Browser is running on. This should also be possible in Firefox. PoC anyone?

There are just so many holes to determine your fingerprint, that we will never patch them all. OpSec peoples .. OpSec

**Comment by crssi** [comment]
Yeah... resistance is futile :(

**Comment by Thorin-Oakenpants** [comment]
Get your own soapbox crssi :D~

**Comment by crssi** [comment]
^^ I am afraid I haven't understand. Maybe my english language (phrases) limitations. Please, enlighten me.

**Comment by Thorin-Oakenpants** [comment]
https://en.wikipedia.org/wiki/Soapbox

> ... as in the phrases "He's on his soapbox", or "Get off your soapbox." ...

hence I said .. "Get your own soapbox". Generally speaking a soapbox

> The term originates from the days when speakers would elevate themselves by standing on a wooden crate originally used for shipment of soap or other dry goods

Each orator would bring their own box .. a soapbox is for one person .. this is my soapbox :) so to speak.

Was just diggin at ya ... of course comments are welcome - I just wanted a sticky where I could post interesting things as I came across them rather than create issues for every single little thing

**Comment by earthlng** [comment]
> CSS line-height reveals the platform Tor Browser is running on. This should also be possible in Firefox. PoC anyone?

There's a public PoC that achieves the same thing and is from the same guy. He apparently has about 12 of these.

copy for people who don't want to allow JS on a hackers website:
https://earthlng.github.io/testpages/OS_detection_scrollbar.html

**Comment by Thorin-Oakenpants** [comment]
^^ That test said I was using TOR .. but I'm not. If he meant TBB, rather than the protocol, then it's still wrong - I'm using FF. I guess he hardcoded that since it was a PoC for TBB? (I can't be arsed looking at the code).

It did get my OS right tho - windows

**Comment by earthlng** [comment]
It's because of privacy.resistFingerprinting=true

**Comment by Thorin-Oakenpants** [comment]
Err what? What is? This PoC includes detection of resistFP? huh?

I thought your test page used JUST css line height to detect OS. Correct? (I do not care about the attempt to detect my browser, that is trivial anyway)

**Comment by earthlng** [comment]
> I thought your test page used JUST css line height to detect OS

No it doesn't. I said there's a PoC that **achieves** the same thing ie OS detection.
This PoC detects OS based on scrollbars. The Tor detection part is looking at some of the width+height values

**Comment by Thorin-Oakenpants** [comment]
Well, where's my PoC on CSS line height

**Comment by Thorin-Oakenpants** [comment]
Also there are already tor tickets for scrollbars, I even brought it up once somewhere (not saying where)

**Comment by Thorin-Oakenpants** [comment]
FYI: https://www.eff.org/deeplinks/2017/08/how-captive-portals-interfere-wireless-security-and-privacy

**Comment by Thorin-Oakenpants** [comment]
Say [Bye Bye Bye](https://en.wikipedia.org/wiki/Bye_Bye_Bye) ... 

[1350646](https://bugzilla.mozilla.org/show_bug.cgi?id=1350646) The clean-up has begun :toilet: 
[1347507](https://bugzilla.mozilla.org/show_bug.cgi?id=1347507) tonnes more to go :

**Comment by Thorin-Oakenpants** [comment]
https://torrentfreak.com/dmca-used-to-remove-ad-server-url-from-easylist-ad-blocklist-170811/

Time to add that sucker to your custom rules...

**Comment by Thorin-Oakenpants** [comment]
https://www.schneier.com/blog/archives/2017/08/confusing_self-.html

AWESOME!!! Anarchists unite! .. err wot?

**Comment by earthlng** [comment]
[Bug 1355576](https://bugzilla.mozilla.org/show_bug.cgi?id=1355576) - Implement clearing of LocalStorage in browsingData API
RESOLVED FIXED in Firefox 57

**Comment by Thorin-Oakenpants** [comment]
^^ I can see a lot of WebExt issues being dealt with before 57->Stable

**Comment by Thorin-Oakenpants** [comment]
OK.. I just lost it ... laughed so hard, I think a little wee came out :smile: 

[Hacking a Gene Sequencer by Encoding Malware in a DNA Strand](https://www.schneier.com/blog/archives/2017/08/hacking_a_gene_.html) ... and someone in thee comments said this

> Cartoon idea: Kid and parent talking to an angry pediatrician. The kid has grotesque appendages growing out of his neck and shoulders. The parent is saying, "Why yes. We call him 'Little Bobby Tables....'"

**Comment by Thorin-Oakenpants** [comment]
[The Book of Mozilla](https://bugzilla.mozilla.org/show_bug.cgi?id=1370613) .. Mammon's new adventures continue

**Comment by Thorin-Oakenpants** [comment]
https://scotthelme.ghost.io/im-giving-up-on-hpkp/

**Comment by earthlng** [comment]
/r/Firefox - [New privacy related preference in Nightly](https://www.reddit.com/r/firefox/comments/6wmpd0/new_privacy_related_preference_in_nightly/)

> We're detecting a long list of obscure and some not so obscure 3rd party Windows applications that use accessibility to snoop on user activity. We want to cut down on this by better communicating when a11y is active.

= `accessibility.force_disabled`

**Comment by Thorin-Oakenpants** [comment]
The new permissions page is long overdue

**Comment by Thorin-Oakenpants** [comment]
> in particular, they changed the initial scheme (moz-extension://[extID]/[path]) to moz-extension://[random-UUID]/[path]. Unfortunately, while this change makes indeed more difficult to enumerate user extensions, it introduces a far more dangerous problem. In fact, the random-UUID token can now be used to precisely fingerprint users if it is leaked by an extensions. A website can retrieve this UUID and use it to uniquely identify the user, as once it is generated the random ID never changes. We reported this design-related bug to Firefox developers as well.

Well, ***that's*** not very good then is it? [Reading for NERDs](https://www.usenix.org/system/files/conference/usenixsecurity17/sec17-sanchez-rola.pdf) (it's a pdf)

**Comment by earthlng** [comment]
> Well, ***that's*** not very good then is it? 

Nope it certainly isn't. Screenshots for example has content-accessible resources but it needs user-interaction to get the UUID. It only took me a couple minutes to create a PoC but I'm not sure whether I should share it or not.
Hopefully they'll get a quick fix done, and hopefully they'll also ship it to ESR.
There rest of their detection method however is not very problematic, at least for us, because it relies on the performance API which is now also covered by RFP.

**Comment by Thorin-Oakenpants** [comment]
Nooooo .. leave the kitty alone.

We may have to vet WebExt for content-accessible resources? Also, if they move to random per "use" (its in the very last section of the pdf), then allowing a cookie exception per addon via FF page info>permissions will break ..

**Comment by earthlng** [comment]
You use screenshots, right? Did you test it?

**Comment by Thorin-Oakenpants** [comment]
I hit screenshots and drew a rectangle .. didn't save anything .. was waiting for a message

**Comment by 2glops** [comment]
I tried screenshots, after drawing my selection I have to wait some seconds with the mouse button on to see the proposals (cancel / download / cloud) icons

**Comment by Thorin-Oakenpants** [comment]
^^ Obviously I allowed JS, I also allowed postimg thru to see the cute kitten. There's no xhr showing in uMatrix. Anyway, As soon you hit the Screenshot button and move your mouse it should tell you underneath in red bold letters your UUID. And the picture changes. No need to actually take a screenshot

**Comment by earthlng** [comment]
There's is no xhr because I'm not interested in your UUID's - it's just a quick and dirty PoC, and I also only tested it with the Screenshots version in nightly 56.

**Comment by Thorin-Oakenpants** [comment]
I logged an issue with github screenshots - so leave that PoC up please

**Comment by 2glops** [comment]
After test with JS enabled, yes it got my screenshots UUID.
Then I remove the UUID from "extensions.webextensions.uuids" in about:config and a new UUID was assigned after a new test.

**Comment by crssi** [comment]
What I am doing wrong? Screenshot with ctrl+shift+M, but nothing happens.

**Comment by Thorin-Oakenpants** [comment]
@crssi - its the system addon Screenshots

**Comment by crssi** [comment]
Now I feel dumb. What system addon Screenshots?

**Comment by Thorin-Oakenpants** [comment]
@crssi - go look in 
     * Portable: "...\App\Firefox64\browser\features\" (or "App\Firefox\etc" for 32bit)
     * Windows: "...\Program Files\Mozilla\browser\features" (or "Program Files (X86)\etc" for 32bit)
     * Mac: "...\Applications\Firefox\Contents\Resources\browser\features\"
and you should have a `screenshots@mozilla.org.xpi`- if not, go grab a portable FF and use that.

Turn on screenshots: 
user_pref("extensions.screenshots.system-disabled", false);
user_pref("extensions.screenshots.disabled", false);

You should get a screenshot toolbar button. See https://www.ghacks.net/2017/05/28/firefox-screenshots-integrated-in-firefox-nightly/ from May for some pretty pictures

**Comment by crssi** [comment]
Thanks @Thorin-Oakenpants 
OK... I can see clearly now... nasty stuff.
Good job @earthlng

**Comment by Thorin-Oakenpants** [comment]
@earthlng .. link .. join the pieces (do not particular wish to link back to here)

https:/github.com/mozilla-services/screenshots/issues/    3429

Ummm .. wots with the link to all this UTM crap. Why do we want UTM all over screenshots?

**Comment by Theemim** [comment]
I wonder if crypto guys are happy with the strength of the UUID stuff.  Some links:

https://dxr.mozilla.org/mozilla-release/source/toolkit/components/extensions/Extension.jsm#188
https://dxr.mozilla.org/mozilla-release/source/xpcom/base/nsIUUIDGenerator.idl
https://dxr.mozilla.org/mozilla-release/source/obj-x86_64-pc-linux-gnu/dist/include/nsIUUIDGenerator.h
https://dxr.mozilla.org/mozilla-release/source/xpcom/base/nsUUIDGenerator.h
https://dxr.mozilla.org/mozilla-release/source/xpcom/base/nsUUIDGenerator.cpp

**Comment by Thorin-Oakenpants** [comment]
The crypto is fine IMO. Look, I'm not Bruce, but I would guess it's fit for purpose (random unique). The problem is the leakage of the UUID

https://bugzilla.mozilla.org/show_bug.cgi?id=1372288  WTF is this? It's tagged "tor fingerprint" and suddenly appeared ("probably something completely different" - _Monty Python_)

**Comment by Thorin-Oakenpants** [comment]
ruh roh... https://www.ghacks.net/2017/08/30/firefox-webextensions-may-identify-you-on-the-internet/ .. watch the FF haters come out - I hope this issue gets some proper attention and/or responses from Mozilla. Hope something happens to 56 (unlikely), but def by 57. Not sure ESR needs it since WE isn;t really a thing in ESR is it?

Not sure why the picture is blurred - shame on u earthlng for using a nude lion :) Interesting that the PoC wasn't linked to.

**Comment by Thorin-Oakenpants** [comment]
Yes .. confirmed, that is the ticket [1372288](https://bugzilla.mozilla.org/show_bug.cgi?id=1372288) - currently access denied, should be opened up soon (no idea if that means days, weeks?) - can't wait for the information on the fix, and timelines - yay! :)

@earthlng - evipie (Tom Shuster) has some info for you - see my link to the screenshots issue a couple of posts up

**Comment by earthlng** [comment]
I thought no more ghacks articles??? I shared my poc in our small little circle here, not for it to be spread far and wide. And now look what it got me ... they're discussing banning ghacks articles on /r/Firefox and they talk about "them" which I assume means Martin and the 2 of us. Not cool, dude! bye bye PoC

**Comment by Atavic** [comment]
/r/Firefox is just one of the many places to discuss.
@earthlng Great PoC

**Comment by Atavic** [comment]
/r/Firefox is just one of the many places for discussions.

**Comment by Thorin-Oakenpants** [comment]
Sorry bout that earthlng.

re ghacks articles - was not decided - I was and still am prepared to see what happens with release 56. Depending on the comments (and its a vocal minority), if it was a bit anti Mozilla then I would not bother with 57, maybe. It's still a very effective medium to get our work across. Why let a few rotten trolls spoil the work. That and Appster who fuels most of the crap, seems to have calmed down.

as for r/firefox - I think its obviously been boiling from the comments that they have had this discussion before. What I think really pissed them off was the PoC images. I mean, I'm still laughing - love it (don't change man). The moderator et al seem pissed off for some reason - surely it can't be that the vulnerability exists, or that its been disclosed (by that pdf research group), or that ~~we~~ you coded a PoC, or that ~~we~~ I showed Martin (I did not know he would do an article). I think some of them are just butthurt TBH. There's nothing about that article that should upset people. It says Safari has the same issue. It says Screenshots requires user intervention. It even says that Mozilla are aware of it. Dunno why some r/firefox are so out of shape over it all.

Soz again for sharing your PoC. I had already pushed the issue at Screenshots and should have just let it ride for a while.

tl;dr: screw r/firefox if they want to act all defensive and moderate - its their bed, they can sleep in it. We're not responsible for anything outside this repo. Martin publishes of his own free will, and we have no control over how he portrays things. Martin is also a very handy tool for us to promote what we do here.

I shall shave my head and meditate on all this for seven years
